netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,313 @@
+      return super.writeData(ctx, streamId, data, padding, endOfStream, promise);
+    }
+    PendingStream pendingStream = pendingStreams.get(streamId);
+    if (pendingStream != null) {
+      pendingStream.frames.add(new DataFrame(data, padding, endOfStream, promise));
+    } else {
+      data.release();
+      promise.setFailure(connectionError(PROTOCOL_ERROR, "Stream does not exist %d", streamId));
+    }
+    return promise;
+  }
+
+  @Override
+  public ChannelFuture writeGoAway(ChannelHandlerContext ctx, int lastStreamId, long errorCode,
+                                   ByteBuf debugData, ChannelPromise promise) {
+    cancelPendingStreams();
+    lastStreamId = min(largestCreatedStreamId, lastStreamId);
+    return super.writeGoAway(ctx, lastStreamId, errorCode, debugData, promise);
+  }
+
+  @Override
+  public ChannelFuture writeSettingsAck(ChannelHandlerContext ctx, ChannelPromise promise) {
+    ChannelFuture future = super.writeSettingsAck(ctx, promise);
+    // After having received a SETTINGS frame, the maximum number of concurrent streams
+    // might have changed. So try to create some buffered streams.
+    tryCreatePendingStreams();
+    return future;
+  }
+
+  @Override
+  public void close() {
+    super.close();
+    cancelPendingStreams();
+  }
+
+  private void tryCreatePendingStreams() {
+    while (!pendingStreams.isEmpty() && connection().local().acceptingNewStreams()) {
+      createNextPendingStream();
+    }
+  }
+
+  private void cancelPendingStreams() {
+    while (!pendingStreams.isEmpty()) {
+      PendingStream stream = pendingStreams.pollFirstEntry().getValue();
+      cancelPendingStream(stream);
+    }
+  }
+
+  private void createNextPendingStream() {
+    Map.Entry<Integer, PendingStream> entry = pendingStreams.pollFirstEntry();
+    if (entry == null) {
+      return;
+    }
+    PendingStream pendingStream = entry.getValue();
+    for (Frame frame : pendingStream.frames) {
+      if (frame instanceof HeadersFrame) {
+        HeadersFrame headers = (HeadersFrame) frame;
+        super.writeHeaders(pendingStream.ctx, pendingStream.streamId, headers.headers,
+                           headers.streamDependency, headers.weight, headers.exclusive,
+                           headers.padding, headers.endOfStream, headers.promise);
+      } else {
+        DataFrame data = (DataFrame) frame;
+        super.writeData(pendingStream.ctx, pendingStream.streamId, data.data, data.padding,
+                        data.endOfStream, data.promise);
+      }
+    }
+    largestCreatedStreamId = pendingStream.streamId;
+  }
+
+  private static void cancelPendingStream(PendingStream stream) {
+    for (Frame frame : stream.frames) {
+      frame.promise.setFailure(connectionError(PROTOCOL_ERROR, "Buffered stream canceled."));
+    }
+  }
+
+  private boolean existingStream(int streamId) {
+    return streamId <= largestCreatedStreamId;
+  }
+
+  private static class PendingStream {
```

Comment:
@ejona86 One "optimisation" I thought of, is having the `PendingStream` extend `HeadersFrame`. This would safe us one object allocation basically. I am not strongly in favor of that one though. WDYT?